pkg/engine/task/task.go

```diff
@@ -36,49 +38,105 @@ const (
 	ApplyTaskKind  = "Apply"
 	DeleteTaskKind = "Delete"
 	DummyTaskKind  = "Dummy"
+	PipeTaskKind   = "Pipe"
 )
 
 var (
-	taskRenderingError   = "TaskRenderingError"
-	taskEnhancementError = "TaskEnhancementError"
-	dummyTaskError       = "DummyTaskError"
+	taskRenderingError      = "TaskRenderingError"
+	taskEnhancementError    = "TaskEnhancementError"
+	dummyTaskError          = "DummyTaskError"
+	resourceUnmarshalError  = "ResourceUnmarshalError"
+	resourceValidationError = "ResourceValidationError"
 )
 
 // Build factory method takes an v1beta1.Task and returns a corresponding Tasker object
 func Build(task *v1beta1.Task) (Tasker, error) {
 	switch task.Kind {
 	case ApplyTaskKind:
-		return newApply(task), nil
+		return newApply(task)
 	case DeleteTaskKind:
-		return newDelete(task), nil
+		return newDelete(task)
 	case DummyTaskKind:
-		return newDummy(task), nil
+		return newDummy(task)
+	case PipeTaskKind:
+		return newPipe(task)
 	default:
 		return nil, fmt.Errorf("unknown task kind %s", task.Kind)
 	}
 }
 
-func newApply(task *v1beta1.Task) ApplyTask {
+func newApply(task *v1beta1.Task) (Tasker, error) {
+	// validate ApplyTask
+	if len(task.Spec.ResourceTaskSpec.Resources) == 0 {
+		return nil, errors.New("task validation error: apply task has an empty resource list. if that's what you need, use a Dummy task instead")
+	}
+
 	return ApplyTask{
 		Name:      task.Name,
 		Resources: task.Spec.ResourceTaskSpec.Resources,
-	}
+	}, nil
 }
 
-func newDelete(task *v1beta1.Task) DeleteTask {
+func newDelete(task *v1beta1.Task) (Tasker, error) {
+	// validate DeleteTask
+	if len(task.Spec.ResourceTaskSpec.Resources) == 0 {
+		return nil, errors.New("task validation error: delete task has an empty resource list. if that's what you need, use a Dummy task instead")
+	}
+
 	return DeleteTask{
 		Name:      task.Name,
 		Resources: task.Spec.ResourceTaskSpec.Resources,
-	}
+	}, nil
 }
 
-func newDummy(task *v1beta1.Task) DummyTask {
+func newDummy(task *v1beta1.Task) (Tasker, error) {
 	return DummyTask{
 		Name:    task.Name,
 		WantErr: task.Spec.DummyTaskSpec.WantErr,
 		Fatal:   task.Spec.DummyTaskSpec.Fatal,
 		Done:    task.Spec.DummyTaskSpec.Done,
+	}, nil
+}
+
+func newPipe(task *v1beta1.Task) (Tasker, error) {
+	// validate PipeTask
+	if len(task.Spec.PipeTaskSpec.Pipe) == 0 {
+		return nil, errors.New("task validation error: pipe task has an empty pipe files list")
+	}
+
+	var pipeFiles []PipeFile
+	for _, pp := range task.Spec.PipeTaskSpec.Pipe {
+		pf := PipeFile{File: pp.File, Kind: pp.Kind, Key: pp.Key}
+		// validate pipe file
+		if err := validPipeFile(pf); err != nil {
+			return nil, err
+		}
+		pipeFiles = append(pipeFiles, pf)
+	}
+
+	return PipeTask{
+		Name:      task.Name,
+		Container: task.Spec.PipeTaskSpec.Container,
+		PipeFiles: pipeFiles,
+	}, nil
+}
+
+var (
+	pipeFileKeyRe  = regexp.MustCompile(`^[a-zA-Z0-9_\-]+$`)    //a-z, A-Z, 0-9, _ and - are allowed
```

Comment:
This is basically a param-name, right? Do we have a specification/pattern for params somewhere else already? Might be good to reuse that, if we have it

Reply:
Not quite. This is k8s limitation for ConfigMap data keys. I've added a comment.

Reply:
As far as I can see we're only testing for [invalid characters](https://github.com/kudobuilder/kudo/blob/master/pkg/kudoctl/cmd/verify/verify.go#L18). Am I being too strict here? 🤔

Reply:
Hmmmm. We should allow all parameter names that K8s allows. Otherwise we might run into a problem where some application expects a parameter name of `"option=rw,size=2300"`. We just had that case in Marathon, where some application abused the parameter name to pass multiple things :-/

In any case, the validation should be the same as is used in the verify that you linked.